tests_integration/test_git.py

```diff
@@ -0,0 +1,271 @@
+    STAGE_NAME = "a_perfect_stage_for_testing"
+    result = runner.invoke_with_connection(["sql", "-q", f"create stage {STAGE_NAME}"])
+    assert result.exit_code == 0
+
+    # copy - test error messages
+    result = runner.invoke_with_connection(
+        ["git", "copy", git_repository, f"@{STAGE_NAME}"]
+    )
+    _assert_error_message_in_output(result.output)
+    try:
+        repository_path = f"@{git_repository}"
+        runner.invoke_with_connection(
+            ["git", "copy", repository_path, f"@{STAGE_NAME}"]
+        )
+    except ProgrammingError as err:
+        assert (
+            err.raw_msg
+            == "Files paths in git repositories must specify a scope. For example, a branch name, "
+            "a tag name, or a valid commit hash. Commit hashes are between 6 and 40 characters long."
+        )
+
+    try:
+        repository_path = f"@{git_repository}/branches/branch_which_does_not_exist"
+        runner.invoke_with_connection(
+            ["git", "copy", repository_path, f"@{STAGE_NAME}"]
+        )
+    except ProgrammingError as err:
+        assert (
+            err.raw_msg
+            == "The specified branch 'branch_which_does_not_exist' cannot be found in the Git Repository."
+        )
+
+    try:
+        repository_path = f"@{git_repository}/tags/tag_which_does_not_exist"
+        runner.invoke_with_connection(
+            ["git", "copy", repository_path, f"@{STAGE_NAME}"]
+        )
+    except ProgrammingError as err:
+        assert (
+            err.raw_msg
+            == "The specified tag 'tag_which_does_not_exist' cannot be found in the Git Repository."
+        )
+
+    # copy to stage - success
+    repository_path = f"@{git_repository}/branches/{BRANCH_NAME}"
+    result = runner.invoke_with_connection_json(
+        ["git", "copy", repository_path, f"@{STAGE_NAME}"]
+    )
+    assert result.exit_code == 0
+    assert result.json == [
+        {"file": "an_existing_directory/file_inside.txt"},
+        {"file": "file.txt"},
+        {"file": "file_only_present_on_a_different_branch.txt"},
+    ]
+
+    result = runner.invoke_with_connection_json(["object", "stage", "list", STAGE_NAME])
+    assert result.exit_code == 0
+    assert _filter_key(result.json, key="name") == [
+        f"{STAGE_NAME}/an_existing_directory/file_inside.txt",
+        f"{STAGE_NAME}/file.txt",
+        f"{STAGE_NAME}/file_only_present_on_a_different_branch.txt",
+    ]
+
+    # copy to local file system - success
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        repository_path = f"@{git_repository}/tags/{TAG_NAME}"
+        result = runner.invoke_with_connection_json(
+            ["git", "copy", repository_path, tmp_dir]
+        )
+        assert result.exit_code == 0
+        assert (Path(tmp_dir) / "file_only_present_on_a_particular_tag.txt").exists()
+        assert _filter_key(result.json, key="file") == [
+            f"an_existing_directory/file_inside.txt",
```

Comment:
This is the result returned by Snowflake but do I correctly understand that there will be no `an_existing_directory` because of flattening issue?

Reply:
That was the case; I modified the test, I'm checking only results